public/scripts/project/lock.js

```diff
@@ -0,0 +1,75 @@
+define(function(require) {
+  var fs = Bramble.getFileSystem();
+  var sh = new fs.Shell();
+  var LOCK_KEY = 'bramble-lock';
+  var EXPIRY = 2000; // MS
+
+  function lock(root, callback) {
+    function refresh(callback) {
+      fs.setxattr(root, LOCK_KEY, Date.now(), function(err) {
+        if (err) {
+          return callback(err);
+        }
+
+        callback();
+      });
+    }
+
+    function setLock() {
+      function handleError(err) {
+        console.error("[Bramble] Failed to refresh project lock");
+      }
+
+      refresh(function(err) {
+        if (err) {
+          return callback(err);
+        }
+
+        setInterval(refresh.bind(null, handleError), EXPIRY);
+        callback();
+      });
+    }
+
+    isLocked(root, function(err) {
+      if (err) {
+        return callback(err);
+      }
+
+      setLock();
+    });
+  }
+
+  function isLocked(root, callback) {
+    // First, make sure the directory exists
+    sh.mkdirp(root, function(err) {
+      if (err && err !== "EEXIST") {
+        return callback(err);
+      }
+
+      // Then we try and access the lock
+      fs.getxattr(root, LOCK_KEY, function(err, value) {
+        if (err) {
+          if (err !== "ENOENT") {
+            return callback(err);
+          }
+
+          // It's never been locked
+          return callback();
+        }
+
+        var threshold = Date.now() - value;
+        if (threshold >= EXPIRY) {
```

Comment:
Should we add a fudge factor here, so that we don't lock out another tab that's trying to reacquire the lock (i.e., give a reacquire attempt a higher chance of success than a new lock attempt)?